src/libostree/ostree-sysroot-deploy.c

```diff
@@ -614,6 +615,83 @@ checkout_deployment_tree (OstreeSysroot *sysroot, OstreeRepo *repo, OstreeDeploy
                                 cancellable, error))
     return FALSE;
 
+#ifdef HAVE_COMPOSEFS
+  if (repo->composefs_wanted != OT_TRISTATE_NO)
+    {
+      static const char integrity_section[] = "ex-integrity";
+      g_autoptr (GBytes) sig = NULL;
+      gboolean apply_composefs_signature;
+      g_autofree guchar *fsverity_digest = NULL;
+      g_auto (GLnxTmpfile) tmpf = {
+        0,
+      };
+      g_autoptr (GVariant) commit_variant = NULL;
+
+      if (!ostree_repo_load_commit (repo, revision, &commit_variant, NULL, error))
+        return FALSE;
+
+      if (!ot_keyfile_get_boolean_with_default (repo->config, integrity_section,
+                                                "composefs-apply-sig", TRUE,
+                                                &apply_composefs_signature, error))
+        return FALSE;
+
+      g_autoptr (GVariant) metadata = g_variant_get_child_value (commit_variant, 0);
+      g_autoptr (GVariant) metadata_composefs
+          = g_variant_lookup_value (metadata, "ostree.composefs", G_VARIANT_TYPE_BYTESTRING);
+      g_autoptr (GVariant) metadata_composefs_sig
+          = g_variant_lookup_value (metadata, "ostree.composefs-sig", G_VARIANT_TYPE_BYTESTRING);
+
+      /* Create a composefs image and put in deploy dir as .ostree.cfs */
+      g_autoptr (OstreeComposefsTarget) target = ostree_composefs_target_new ();
+
+      g_autoptr (GFile) commit_root = NULL;
+      if (!ostree_repo_read_commit (repo, csum, &commit_root, NULL, cancellable, error))
+        return FALSE;
+
+      if (!ostree_repo_checkout_composefs (repo, target, (OstreeRepoFile *)commit_root, cancellable,
+                                           error))
+        return FALSE;
+
+      g_autofree char *composefs_cfs_path
+          = g_strdup_printf ("%s/.ostree.cfs", checkout_target_name);
+
+      if (!glnx_open_tmpfile_linkable_at (osdeploy_dfd, checkout_target_name, O_WRONLY | O_CLOEXEC,
+                                          &tmpf, error))
+        return FALSE;
+
+      if (!ostree_composefs_target_write (target, tmpf.fd, &fsverity_digest, cancellable, error))
+        return FALSE;
+
+      /* If the commit specified a composefs digest, verify it */
+      if (metadata_composefs
+          && g_variant_n_children (metadata_composefs) == OSTREE_SHA256_DIGEST_LEN
+          && memcmp (fsverity_digest, g_variant_get_data (metadata_composefs),
+                     OSTREE_SHA256_DIGEST_LEN)
+                 != 0)
+        return glnx_throw (error, "Generated composefs image digest doesn't match commit metadata");
```

Comment:
It'd be a bit friendlier to output the mismatched hex digests potentially?  Also, with stuff like this I think it's worth having a helper function to compare digests because it's security relevant.  See e.g. `_ostree_compare_object_checksum()`.  And the logic for what to do if the length of the input byte array doesn't match would probably be there.